keel-core/src/main/kotlin/com/netflix/spinnaker/keel/services/ApplicationService.kt

```diff
@@ -443,15 +461,45 @@ class ApplicationService(
     }
   }
 
+  private fun getArtifactVersionByPromotionStatus(
+    environmentName: String,
+    promotionStatus: PromotionStatus,
+    version: String?,
+    artifactInfoInEnvironment: Map<String, List<StatusInfoForArtifactInEnvironment>>,
+    allVersions: List<PublishedArtifact>
+  ): PublishedArtifact? {
+    //only CURRENT and PREVIOUS are supported, as they can be sorted by deploy_at
+    require(promotionStatus in listOf(CURRENT, PREVIOUS)) { "Invalid promotion status used to query" }
+    val currentEnvVersionInfo = artifactInfoInEnvironment[environmentName] ?: return null
+    // we sort the versions by their deployed at time in descending order, then we take
+    // the first one where all the conditions match.
+    val chosenVersion = if (version == null) {
+      currentEnvVersionInfo.sortedByDescending { it.deployedAt }.firstOrNull { it.status == promotionStatus }
+    } else {
+      currentEnvVersionInfo.sortedByDescending { it.deployedAt }.firstOrNull { it.status == promotionStatus && it.replacedByVersion == version}
+    } ?: return null
+    return allVersions.find { it.version == chosenVersion.version }
```

Comment:
Might be worth checking if binary search gives us a performance boost here (although we would need to pass in the sorting strategy).